src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,410 @@
+                    assert(!"can't restore signature argument value");
+                    continue;
+                }
+
+                // Restore runtimeLookup using signature argument via a global dictionary
+                CORINFO_RUNTIME_LOOKUP runtimeLookup = {};
+                const bool             lookupFound   = GetSignatureToLookupInfoMap()->Lookup(signature, &runtimeLookup);
+                assert(lookupFound);
+
+                const bool needsSizeCheck = runtimeLookup.sizeOffset != CORINFO_NO_SIZE_CHECK;
+                if (needsSizeCheck)
+                {
+                    JITDUMP("dynamic expansion, needs size check.\n")
+                }
+
+                DebugInfo debugInfo = stmt->GetDebugInfo();
+
+                assert(runtimeLookup.indirections != 0);
+                assert(runtimeLookup.testForNull);
+
+                // Split block right before the call tree
+                BasicBlock* prevBb       = block;
+                GenTree**   callUse      = nullptr;
+                Statement*  newFirstStmt = nullptr;
+                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+                assert(prevBb != nullptr && block != nullptr);
+
+                GenTreeLclVar* rtLookupLcl = nullptr;
+
+                // Mostly for Tier0: if the current statement is ASG(LCL, RuntimeLookup)
+                // we can drop it and use that LCL as the destination
+                if (stmt->GetRootNode()->OperIs(GT_ASG))
+                {
+                    GenTree* lhs = stmt->GetRootNode()->gtGetOp1();
+                    GenTree* rhs = stmt->GetRootNode()->gtGetOp2();
+                    if (lhs->OperIs(GT_LCL_VAR) && rhs == *callUse)
+                    {
+                        rtLookupLcl = gtClone(lhs)->AsLclVar();
+                        fgRemoveStmt(block, stmt);
+                    }
+                }
+
+                // Grab a temp to store result (it's assigned from either fastPathBb or fallbackBb)
+                if (rtLookupLcl == nullptr)
+                {
+                    // Define a local for the result
+                    unsigned rtLookupLclNum         = lvaGrabTemp(true DEBUGARG("runtime lookup"));
+                    lvaTable[rtLookupLclNum].lvType = TYP_I_IMPL;
+                    rtLookupLcl                     = gtNewLclvNode(rtLookupLclNum, call->TypeGet());
+
+                    // Replace call with rtLookupLclNum local and update side effects
+                    *callUse = gtClone(rtLookupLcl);
+                    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+                    {
+                        fgMorphStmtBlockOps(block, newFirstStmt);
+                        newFirstStmt = newFirstStmt->GetNextStmt();
+                    }
+                    fgMorphStmtBlockOps(block, stmt);
+                    gtUpdateStmtSideEffects(stmt);
+                }
```

Comment:
```suggestion
                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
                assert(prevBb != nullptr && block != nullptr);

                // Block ops inserted by the split need to be morphed here since we are after morph.
                // We cannot morph stmt yet as we may modify it further below, and the morphing
                // could invalidate callUse.
                while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
                {
                    fgMorphStmtBlockOps(block, newFirstStmt);
                    newFirstStmt = newFirstStmt->GetNextStmt();
                }

                GenTreeLclVar* rtLookupLcl = nullptr;

                // Mostly for Tier0: if the current statement is ASG(LCL, RuntimeLookup)
                // we can drop it and use that LCL as the destination
                if (stmt->GetRootNode()->OperIs(GT_ASG))
                {
                    GenTree* lhs = stmt->GetRootNode()->gtGetOp1();
                    GenTree* rhs = stmt->GetRootNode()->gtGetOp2();
                    if (lhs->OperIs(GT_LCL_VAR) && rhs == *callUse)
                    {
                        rtLookupLcl = gtClone(lhs)->AsLclVar();
                        fgRemoveStmt(block, stmt);
                    }
                }

                // Grab a temp to store result (it's assigned from either fastPathBb or fallbackBb)
                if (rtLookupLcl == nullptr)
                {
                    // Define a local for the result
                    unsigned rtLookupLclNum         = lvaGrabTemp(true DEBUGARG("runtime lookup"));
                    lvaTable[rtLookupLclNum].lvType = TYP_I_IMPL;
                    rtLookupLcl                     = gtNewLclvNode(rtLookupLclNum, call->TypeGet());

                    *callUse = gtClone(rtLookupLcl);

                    fgMorphStmtBlockOps(block, stmt);
                    gtUpdateStmtSideEffects(stmt);
                }
```

Otherwise we would fail to morph the block ops in the inserted statements by `gtSplitTree`.